client/src/main/scala/org/http4s/client/ClientFailure.scala

```diff
@@ -0,0 +1,12 @@
+package org.http4s.client
+
+import java.io.IOException
+import java.net.InetSocketAddress
+
+/** Indicates a failure in the client when executing a request preserving the context (requestKey and upstream address)
+  */
+class ClientFailure(requestKey: RequestKey, upstream: InetSocketAddress, cause: Throwable)
```

Comment:
Maybe we should make `requestKey` and `upstream` vals in case anyone wants to do something structurally.

I don't like making exceptions case classes, because that implies equality if the arguments are equal, and I tend not to think two exceptions are equal if their arguments are.  But we might even consider an `unapply` on a companion object.

Reply:
makes sense, i will do that later today